test/deployframework/framework.go

```diff
@@ -0,0 +1,335 @@
+	NamespacePrefix            string
+	ManifestsDir               string
+	LoggingPath                string
+	CleanupScriptPath          string
+	KubeConfigPath             string
+	ReportingOperatorImageRepo string
+	ReportingOperatorImageTag  string
+	Logger                     logrus.FieldLogger
+	Client                     kubernetes.Interface
+	APIExtClient               apiextclientv1beta1.CustomResourceDefinitionsGetter
+	MeteringClient             meteringv1.MeteringV1Interface
+	Deployer                   *deploy.Deployer
+	Config                     ReportingFrameworkConfig
+}
+
+// ReportingFrameworkConfig is a structure containing information
+// needed to customize a ReportingFramework object
+type ReportingFrameworkConfig struct {
+	UseKubeProxyForReportingAPI bool
+	UseRouteForReportingAPI     bool
+	HTTPSAPI                    bool
+	Namespace                   string
+	KubeConfigPath              string
+	ReportingAPIURL             string
+	RouteBearerToken            string
+	ReportResultsOutputPath     string
+}
+
+// New is the constructor function that creates and returns a new DeployFramework object
+func New(
+	logger logrus.FieldLogger,
+	nsPrefix,
+	manifestDir,
+	kubeconfig,
+	cleanupScriptPath,
+	loggingPath,
+	reportingOperatorImageRepo,
+	reportingOperatorImageTag string,
+) (*DeployFramework, error) {
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build a kube config from %s: %v", kubeconfig, err)
+	}
+
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize the k8s clientset: %v", err)
+	}
+
+	apiextClient, err := apiextclientv1beta1.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize the apiextensions clientset: %v", err)
+	}
+
+	meteringClient, err := meteringv1.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize the metering clientset: %v", err)
+	}
+
+	deployFramework := &DeployFramework{
+		NamespacePrefix:            nsPrefix,
+		ManifestsDir:               manifestDir,
+		CleanupScriptPath:          cleanupScriptPath,
+		KubeConfigPath:             kubeconfig,
+		LoggingPath:                loggingPath,
+		ReportingOperatorImageRepo: reportingOperatorImageRepo,
+		ReportingOperatorImageTag:  reportingOperatorImageTag,
+		Logger:                     logger,
+		Client:                     client,
+		APIExtClient:               apiextClient,
+		MeteringClient:             meteringClient,
+		Config:                     ReportingFrameworkConfig{},
+	}
+
+	return deployFramework, nil
+}
+
+// Setup handles the process of deploying metering, and waiting for all necessary resources
+// to become ready in order to proceed with running the reporting tests.
+func (df *DeployFramework) Setup(cfg deploy.Config, namespace, testOutputPath string, targetPods int) (*ReportingFrameworkConfig, error) {
+	var err error
+
```

Comment:
There's a lot of mutation of `df` happening during `Setup` that I think could be avoided by doing more of the setup during `New`.  Anything assigning to `df` or `cfg` in here I would consider something which we want to avoid in general if we can. 

Partially initialized objects is one issue, as it can be harder to test and understand in general. Additionally, this makes it possible to use incorrectly, as you could call teardown before setup for example, and teardown might be expecting some of these values to have been set during `Setup`. This might point to the idea that `Setup` should return something we can use for Teardown, but I'm not 100% convinced we need that.

Thing I think we can do:
- Can we try to initialize the `DeployFramework` as much as possible before `Setup` or `Teardown` is called?
- Avoid assigning to `deploy.Config` during setup. Instead I think it would be best if `deploy.Config` was automatically created within `Setup`, not passed as an argument. 90% of the values for `deploy.Config` do not need to be specified on a per-test case basis, only MeteringConfig does, and not even that, we only actually need MeteringConfig spec (this way we're not dealing with the namespace being set/unset, we just create the MeteringConfig with it set). So perhaps you should be creating a MeteringConfig given a MeteringConfigSpec, and then creating a DeployConfig using the MeteringConfig object you create.

Reply:
Note, if we remove `deploy.Config` as an argument, we will also want to take the reporting/metering operator image repo/tag values as arguments so that we can override those still, and we'll use those to create the `deploy.Config` within Setup().